python/test/unit/fem/test_petsc_nonlinear_assembler.py

```diff
@@ -602,86 +433,53 @@ def initial_guess_p(x):
-                        (p.function_space.dofmap.index_map, p.function_space.dofmap.index_map_bs),
-                    ],
-                )
-            x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
 
-            snes.solve(None, x)
-            assert snes.getConvergedReason() > 0
-            snes.destroy()
-            Jnorm = Jmat.norm()
-            Fnorm = Fvec.norm()
+            snes_options = {
+                "snes_rtol": 1.0e-15,
+                "snes_max_it": 10,
+                "snes_monitor": None,
+                "ksp_type": "minres",
+            }
+            solver = dolfinx.nls.petsc.SNESSolver(
+                F, [u, p], bcs=bcs, P=P, assembly_type=dolfinx.fem.AssemblyType.block
+            )
+            solver.set_options(snes_options)
+            x, converged_reason, _ = solver.solve()
+            assert converged_reason > 0
+            solver.replace_solution([u, p], x)
+            Jnorm = solver.snes.getJacobian()[0].norm()
+            Fnorm = solver.snes.getFunction()[0].norm()
             xnorm = x.norm()
-            Jmat.destroy()
-            Fvec.destroy()
-            x.destroy()
             return Jnorm, Fnorm, xnorm
 
         def nested():
             """Blocked and nested"""
-            Jmat = create_matrix_nest(J)
-            Pmat = create_matrix_nest(P)
-            Fvec = create_vector_nest(F)
-
-            snes = PETSc.SNES().create(MPI.COMM_WORLD)
-            snes.setTolerances(rtol=1.0e-15, max_it=20)
-            nested_IS = Jmat.getNestISs()
-            snes.getKSP().setType("minres")
-            snes.getKSP().setTolerances(rtol=1e-8)
-            snes.getKSP().getPC().setType("fieldsplit")
-            snes.getKSP().getPC().setFieldSplitIS(["u", nested_IS[0][0]], ["p", nested_IS[1][1]])
-
-            problem = NonlinearPDE_SNESProblem(F, J, [u, p], bcs, P=P)
-            snes.setFunction(problem.F_nest, Fvec)
-            snes.setJacobian(problem.J_nest, J=Jmat, P=Pmat)
-
             u.interpolate(initial_guess_u)
             p.interpolate(initial_guess_p)
-            x = create_vector_nest(F)
-            for x1_soln_pair in zip(x.getNestSubVecs(), (u, p)):
-                x1_sub, soln_sub = x1_soln_pair
-                soln_sub.x.petsc_vec.ghostUpdate(
-                    addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD
-                )
-                soln_sub.x.petsc_vec.copy(result=x1_sub)
-                x1_sub.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
 
-            x.set(0.0)
-            snes.solve(None, x)
-            assert snes.getConvergedReason() > 0
-            snes.destroy()
-            Jnorm = nest_matrix_norm(Jmat)
-            Fnorm = Fvec.norm()
+            solver = dolfinx.nls.petsc.SNESSolver(
+                F, [u, p], J=J, bcs=bcs, assembly_type=dolfinx.fem.AssemblyType.nest, P=P
+            )
+            nested_IS = solver.snes.getJacobian()[0].getNestISs()
```

Comment:
This is an advanced example, so the user would probably go with the direct SNES interface.